src/samples/http/client.c

```diff
@@ -0,0 +1,202 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "sol-mainloop.h"
+#include "sol-http.h"
+#include "sol-http-client.h"
+
+static void
+response_cb(void *userdata, const struct sol_http_client_connection *connection,
+    struct sol_http_response *response)
+{
+    bool verbose = (bool)(long)userdata;
+
+    if (response->response_code != SOL_HTTP_STATUS_OK) {
+        fprintf(stderr, "Finished with error, response code: %d\n", response->response_code);
+        goto err;
+    }
+
+    if (verbose) {
+        uint16_t idx;
+        struct sol_http_param_value *value;
+
+        SOL_HTTP_PARAMS_FOREACH_IDX (&response->param, value, idx) {
+            switch (value->type) {
+            case SOL_HTTP_PARAM_COOKIE:
+                fprintf(stdout, "[COOKIE] %.*s : %.*s\n",
+                    SOL_STR_SLICE_PRINT(value->value.key_value.key),
+                    SOL_STR_SLICE_PRINT(value->value.key_value.value));
+                break;
+            case SOL_HTTP_PARAM_HEADER:
+                fprintf(stdout, "[HEADER] %.*s : %.*s\n",
+                    SOL_STR_SLICE_PRINT(value->value.key_value.key),
+                    SOL_STR_SLICE_PRINT(value->value.key_value.value));
+            default:
+                break;
+            }
+        }
+    }
+
+    fprintf(stdout, "%.*s\n", (int)response->content.used, (char *)response->content.data);
+    sol_quit_with_code(EXIT_SUCCESS);
+
+    return;
+
+err:
+    sol_quit_with_code(EXIT_FAILURE);
+}
+
+static int
+create_post_field_params(struct sol_http_params *params, const char *value)
+{
+    char *aux, *copy = strdup(value);
+
+    if (!copy) {
+        fprintf(stderr, "ERROR: Could not duplicate the string: %s\n", value);
+        return -ENOMEM;
+    }
+
+    aux = strtok(copy, "&");
+    do {
+        size_t len;
+        char *ptr, *key, *val;
+
+        key = aux;
+        len = strlen(key);
+
+        ptr = memmem(key, len, "=", 1);
+        if (!ptr) {
+            fprintf(stderr, "Invalid key/value in %s\n", value);
+            goto err;
+        }
+
+        *ptr = '\0';
+        val = ++ptr;
+
+        if (!sol_http_param_add_copy(params,
+            SOL_HTTP_REQUEST_PARAM_POST_FIELD(key, val))) {
```

Comment:
yep, by default thi is what is done by curl, one must explictly call --data-urlencoded if want to pass data encoded.